<a href="https://colab.research.google.com/github/davidheram/Logistic-regression-proyect/blob/main/Logistic_regression_proyect_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [140]:

!pip install tableone
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np



In [141]:
df = pd.read_csv("https://storage.googleapis.com/breathecode/project-files/bank-marketing-campaign-data.csv", sep=";")

In [142]:
df.shape

(41188, 21)

In [143]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

In [144]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,41188.0,40.024060,10.421250,17.000,32.000,38.000,47.000,98.000
duration,41188.0,258.285010,259.279249,0.000,102.000,180.000,319.000,4918.000
campaign,41188.0,2.567593,2.770014,1.000,1.000,2.000,3.000,56.000
pdays,41188.0,962.475454,186.910907,0.000,999.000,999.000,999.000,999.000
previous,41188.0,0.172963,0.494901,0.000,0.000,0.000,0.000,7.000
emp.var.rate,41188.0,0.081886,1.570960,-3.400,-1.800,1.100,1.400,1.400
cons.price.idx,41188.0,93.575664,0.578840,92.201,93.075,93.749,93.994,94.767
cons.conf.idx,41188.0,-40.502600,4.628198,-50.800,-42.700,-41.800,-36.400,-26.900
euribor3m,41188.0,3.621291,1.734447,0.634,1.344,4.857,4.961,5.045
nr.employed,41188.0,5167.035911,72.251528,4963.600,5099.100,5191.000,5228.100,5228.100


In [145]:
df["y"].value_counts(normalize=True)

,proportion
y,
no,0.887346
yes,0.112654


In [146]:
df["y"] = np.where(df["y"]=="yes",1,0)
print(df['y'].value_counts())

y
0    36548
1     4640
Name: count, dtype: int64


In [147]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

In [148]:
categorical = ["job","marital", "education","month"]
numerical = ["age","duration","euribor3m"]

X = df[categorical + numerical]
y = df["y"]

table_df=df[categorical + numerical + ["y"]].copy()

In [149]:
df[numerical].value_counts()

age  duration  euribor3m
29   117       4.120        4
31   92        4.963        4
30   111       4.963        4
33   64        4.964        4
42   99        4.857        4
                           ..
34   250       4.864        1
               4.962        1
     251       0.654        1
               0.793        1
     248       4.964        1
Name: count, Length: 39560, dtype: int64

In [150]:
from tableone  import TableOne, load_dataset
from scipy import stats
mytable = TableOne(table_df,categorical=categorical, groupby= "y", pval=True)
print(mytable)

                                         Grouped by y                                                     
                                              Missing        Overall              0              1 P-Value
n                                                              41188          36548           4640        
job, n (%)           admin.                             10422 (25.3)    9070 (24.8)    1352 (29.1)  <0.001
                     blue-collar                         9254 (22.5)    8616 (23.6)     638 (13.8)        
                     entrepreneur                         1456 (3.5)     1332 (3.6)      124 (2.7)        
                     housemaid                            1060 (2.6)      954 (2.6)      106 (2.3)        
                     management                           2924 (7.1)     2596 (7.1)      328 (7.1)        
                     retired                              1720 (4.2)     1286 (3.5)      434 (9.4)        
                     self-employed   

In [151]:
df[numerical].corr()

,age,duration,euribor3m
age,1.000000,-0.000866,0.010767
duration,-0.000866,1.000000,-0.032897
euribor3m,0.010767,-0.032897,1.000000


In [152]:
df.groupby("y")[numerical].mean()

,age,duration,euribor3m
y,,,
0,39.911185,220.844807,3.811491
1,40.913147,553.191164,2.123135


In [153]:
for col in categorical:
    print(f"{col}")
    print(pd.crosstab(df[col], df["y"], normalize="index"))
    print("-"*30)

job
y                     0         1
job                              
admin.         0.870274  0.129726
blue-collar    0.931057  0.068943
entrepreneur   0.914835  0.085165
housemaid      0.900000  0.100000
management     0.887825  0.112175
retired        0.747674  0.252326
self-employed  0.895144  0.104856
services       0.918619  0.081381
student        0.685714  0.314286
technician     0.891740  0.108260
unemployed     0.857988  0.142012
unknown        0.887879  0.112121
------------------------------
marital
y                0         1
marital                     
divorced  0.896791  0.103209
married   0.898427  0.101573
single    0.859959  0.140041
unknown   0.850000  0.150000
------------------------------
education
y                           0         1
education                              
basic.4y             0.897510  0.102490
basic.6y             0.917976  0.082024
basic.9y             0.921754  0.078246
high.school          0.891645  0.108355
illiterate           0.777

In [154]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import scale
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import KFold
from sklearn.model_selection import RepeatedStratifiedKFold


In [155]:
scaler = StandardScaler()
pd.DataFrame(scaler.fit_transform(df[numerical]),columns=scaler.get_feature_names_out())

,age,duration,euribor3m
0,1.533034,0.010471,0.712460
1,1.628993,-0.421501,0.712460
2,-0.290186,-0.124520,0.712460
3,-0.002309,-0.413787,0.712460
4,1.533034,0.187888,0.712460
...,...,...,...
41183,3.164336,0.292025,-1.495186
41184,0.573445,0.481012,-1.495186
41185,1.533034,-0.267225,-1.495186
41186,0.381527,0.708569,-1.495186


In [156]:
ohe = OneHotEncoder(handle_unknown='ignore', drop='first')
pd.DataFrame(ohe.fit_transform(df[categorical]).toarray(), columns=ohe.get_feature_names_out())

,job_blue-collar,job_entrepreneur,job_housemaid,job_management,job_retired,job_self-employed,job_services,job_student,job_technician,job_unemployed,...,education_unknown,month_aug,month_dec,month_jul,month_jun,month_mar,month_may,month_nov,month_oct,month_sep
0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0
1,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0
2,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0
3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0
4,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
41183,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
41184,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
41185,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
41186,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0


In [157]:
def scaler(X_train_, X_test_, nums):
  X_train = X_train_.copy()
  X_test = X_test_.copy()
  scaler = StandardScaler()
  # Train
  X_train_scaled = pd.DataFrame(
             scaler.fit_transform(X_train[nums]),
             columns=scaler.get_feature_names_out(),
             index = X_train.index)
  X_train_scaled = X_train_scaled.join(X_train[list(set(X_train.columns)  - set(nums))])
  # Test
  X_test_scaled = pd.DataFrame(
      scaler.transform(X_test[nums]),
      columns = scaler.get_feature_names_out(),
      index = X_test.index)
  X_test_scaled = X_test_scaled.join(X_test[list(set(X_test.columns)  - set(nums))])
  X_test_scaled = X_test_scaled[X_train_scaled.columns]
  return X_train_scaled, X_test_scaled


def to_binary(X_train_, X_test_, cats):
  X_train =  X_train_.copy()
  X_test = X_test_.copy()
  ohe = OneHotEncoder(handle_unknown='ignore', drop='first')
  # train
  X_train_bin = pd.DataFrame(ohe.fit_transform(
      X_train[cats]).toarray(),
      columns = ohe.get_feature_names_out(),
      index = X_train.index)
  X_train_bin = X_train_bin.join(X_train[list(set(X_train.columns)  - set(cats))])
  # test
  X_test_bin = pd.DataFrame(ohe.transform(X_test[cats]).toarray(),
      columns = ohe.get_feature_names_out(),
      index = X_test.index)
  X_test_bin = X_test_bin.join(X_test[list(set(X_test.columns)  - set(cats))])
  X_test_bin = X_test_bin[X_train_bin.columns]
  return X_train_bin, X_test_bin

In [158]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=123)
X_train_ohe, X_test_ohe = to_binary(X_train, X_test, categorical)
X_train_ml, X_test_ml = scaler(X_train_ohe, X_test_ohe, numerical)

In [159]:
X_train_ml

,age,duration,euribor3m,job_unknown,month_may,education_unknown,job_management,job_housemaid,education_illiterate,job_entrepreneur,...,education_basic.6y,education_basic.9y,job_retired,job_services,marital_unknown,marital_single,job_technician,education_high.school,month_jul,education_professional.course
5402,-0.481846,-0.310570,0.707515,0.0,1.0,0.0,0.0,0.0,0.0,1.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0
40128,-1.730916,-0.009892,-1.616536,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,1.0,0.0,0.0,1.0,1.0
11388,-0.193599,-0.834829,0.766454,0.0,0.0,1.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
16361,-1.442669,0.707109,0.768765,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0,1.0,0.0
23389,-0.289682,-0.584264,0.769343,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7763,-1.346587,-0.719183,0.712138,0.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0
15377,0.286812,-0.125537,0.765298,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
17730,1.343718,-0.156376,0.767610,0.0,0.0,0.0,1.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0
28030,-0.097517,1.308464,-1.242099,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,0.0,0.0


In [160]:
model = LogisticRegression()
model.fit(X_train_ml, y_train)
preds = model.predict(X_test_ml)

In [161]:
print(classification_report(preds, y_test))

              precision    recall  f1-score   support

           0       0.98      0.92      0.95      7743
           1       0.34      0.65      0.45       495

    accuracy                           0.90      8238
   macro avg       0.66      0.79      0.70      8238
weighted avg       0.94      0.90      0.92      8238



In [162]:
def grid_lr(X_train, y_train):
    model = LogisticRegression(random_state=666, max_iter=2000)
    class_weight =  [{0:0.01, 1:0.9},
                     {0:0., 1:0.5},
                     "balanced"]
    solvers = ['liblinear', "lbfgs"]
    penalty = ['l2']
    c_values = [  1.0, 0.1, 0.01, 0.001 ]
    grid = dict(solver=solvers,penalty=penalty,C=c_values, class_weight= class_weight)
    cv = RepeatedStratifiedKFold(n_splits=5, n_repeats=2, random_state=1)
    grid_search = GridSearchCV(estimator=model, param_grid=grid, n_jobs=-1, cv=cv,
                           scoring='f1',
                           error_score='raise')
    grid_result = grid_search.fit(X_train, y_train)
    return  grid_result.best_estimator_

In [163]:
best_model = grid_lr(X_train_ml, y_train)
preds = best_model.predict(X_test_ml)
print(classification_report(y_test, preds))

              precision    recall  f1-score   support

           0       0.98      0.85      0.91      7298
           1       0.43      0.89      0.58       940

    accuracy                           0.85      8238
   macro avg       0.71      0.87      0.74      8238
weighted avg       0.92      0.85      0.87      8238



1. Conclusiones del Análisis Exploratorio (EDA)
El Factor Tiempo (Duration): Es la variable con mayor poder de separación. Los clientes que aceptan el depósito tienen llamadas que duran, en promedio, 553 segundos, frente a los 220 segundos de quienes rechazan. Esto indica que la retención telefónica es el precursor más claro del éxito comercial.

Sensibilidad al Entorno Económico: Existe una relación inversa clara con la tasa Euribor 3 meses. El éxito se duplica cuando las tasas son bajas (promedio de 2.12) en comparación con tasas altas (3.81). El producto es más atractivo en climas de menor interés bancario.

Nichos de Oportunidad Categóricos: * Perfil Social: Los estudiantes (31.4%) y jubilados (25.2%) muestran tasas de conversión muy superiores a la media.

Estacionalidad: Los meses de Marzo, Septiembre, Octubre y Diciembre presentan tasas de éxito cercanas al 50%, a pesar de no ser los meses con más llamadas. Mayo, aunque es el mes más activo, es el menos eficiente (6.4% de éxito).

Calidad del Dato: Se identificaron 2,963 outliers en la duración de las llamadas y 469 en la edad. Esto sugiere que el modelo debe ser robusto a valores extremos o que estos casos requieren un tratamiento especial.



2. Conclusiones del Entrenamiento y Modelo
Desempeño del Modelo Final: Tras el ajuste de hiperparámetros (Grid Search), el modelo alcanzó un F1-score de 0.58 para la clase "yes". Logramos un equilibrio notable con una Precisión del 43% y un Recall del 89%.

Interpretación de Métricas: * El Recall (0.89) es la victoria técnica aquí: significa que el modelo captura casi el 90% de los clientes interesados.

La Precisión (0.43) indica que, aunque todavía hay "falsos positivos", hemos reducido drásticamente el ruido en comparación con el modelo inicial que solo tenía un 24% de precisión.

Eficacia de la Regularización: El uso de StandardScaler y la búsqueda de pesos de clase (class_weight) fue vital para que el modelo no ignorara a la minoría (los que sí compran), permitiendo que el algoritmo aprenda los patrones de éxito a pesar del fuerte desbalance de los datos.